Improve rosdep: 1) Friendlier to beginners, 2) Clarifying rosdep is ROS-agnostic (#3719)
ros2/ros2_documentation: source/Tutorials/Intermediate/Rosdep.rst

    Tutorials/Rosdep

Managing Dependencies with rosdep
=================================

.. contents:: Table of Contents
   :depth: 2
   :local:

**Goal:** Manage external dependencies using ``rosdep``.

**Tutorial level:** Intermediate

**Time:** 5 minutes

This tutorial will explain how to manage external dependencies using ``rosdep``.

> What is rosdep?

Reviewer: Consider adding a *why* as well. The motivation for `rosdep` isn't always clear, leading to questions such as [why rosdep instead of others?](https://answers.ros.org/question/215059) and [When should I use rosdep and when should I use apt or pip?](https://answers.ros.org/question/287544)

Reviewer 2: I've added a couple of sentences about this in fcdc39ddd74fa8a73f6fbbaca1a71d2aff19712c .  Let me know what you think.

---------------

``rosdep`` is a dependency management utility that can work with packages and external libraries.
It is a command-line utility for identifying and installing dependencies to build or install a package.
It can be or is invoked:

- Before building a workspace and needing appropriate dependencies to build the packages within

> - When installing packages (e.g. ``sudo apt install ros-{DISTRO}-demo-nodes-cpp``) to check the dependencies needed for it to execute.

Reviewer: This is somewhat confusing: if installing a package using `apt`, `rosdep` would not be used, right?

Reviewer 2: Yes, that's true.  This is pre-existing, but I've reworked this in fcdc39ddd74fa8a73f6fbbaca1a71d2aff19712c .

    While the name suggests it is for ROS, ``rosdep`` is semi-agnostic to ROS.
    You can utilize this powerful tool in non-ROS software projects by installing it as a standalone Python package.
    Successfully running ``rosdep`` relies on ``rosdep keys`` to be available, which can be downloaded from a public git repository with a few simple commands.

A little about package.xml files
--------------------------------

The ``package.xml`` is the file in your software where ``rosdep`` finds the set of dependencies.
It is important that the list of dependencies in the ``package.xml`` is complete and correct, which allows all of the tooling to determine the packages dependencies.
Missing or incorrect dependencies can lead to users not being able to use your package, to packages in a workspace being built out-of-order, and to packages not being able to be released.

The dependencies in the ``package.xml`` file are generally referred to as "rosdep keys".
These dependencies are manually populated in the ``package.xml`` file by the package's creators and should be an exhaustive list of any non-builtin libraries and packages it requires.

These are represented in the following tags:

> ``<depend>``

Reviewer: Perhaps mention/link [REP-149](https://ros.org/reps/rep-0149.html) (and/or some of the others)?

Reviewer 2: Good call, added a link to REP-149 in fcdc39ddd74fa8a73f6fbbaca1a71d2aff19712c.

Firstly, if this is the first time using ``rosdep``, it must be initialized via:

.. code-block:: bash

    sudo rosdep init
    rosdep update

This will initialize rosdep and ``update`` will update the locally cached rosdistro index.
It is a good idea to ``update`` rosdep on occasion to get the latest index.

Finally, we can run ``rosdep install`` to install dependencies.
Typically, this is run over a workspace with many packages in a single call to install all dependencies.
A call for that would appear as the following, if in the root of the workspace with directory ``src`` containing source code.

.. code-block:: bash

    rosdep install --from-paths src -y --ignore-src

Breaking that down:

- ``--from-paths src`` specifies the path to check for ``package.xml`` files to resolve keys for
- ``-y`` means to default yes to all prompts from the package manager to install without prompts
- ``--ignore-src`` means to ignore installing dependencies, even if a rosdep key exists, if the package itself is also in the workspace.

There are additional arguments and options available.
Use ``rosdep -h`` to see them.

> Managing Dependencies with rosdep

Author: Whoops this line must be at the top as it makes the headline in the left pane in the rendered website.

What is rosdep?
---------------

``rosdep`` is dependency management utility that can work with packages and external libraries.
``rosdep`` is a command-line utility for identifying and installing dependencies to build or install a package.
It can be or is invoked when:

- Before building a workspace and needing appropriate dependencies to build the packages within
- Install packages (e.g. ``sudo apt install ros-{DISTRO}-demo-nodes-cpp``) to check the dependencies needed for it to execute
- and more!

It has the ability to work over a single package or over a directory of packages (e.g. workspace).

> While the name may suggest so, ``rosdep`` is semi-agnostic to ROS. You can utilize this powerful tool in non-ROS software project [1]_.

Reviewer 3: ```suggestion
While the name may suggest so, ``rosdep`` is semi-agnostic to ROS.
You can utilize this powerful tool in non-ROS software project [1]_.
```

Author: Looks like just adding a new line was a suggestion? No strong opinion so applied.

What is rosdep?
---------------

> ``rosdep`` is dependency management utility that can work with packages and external libraries.

Reviewer 2: While this is true, I also think that we should be clear that it is most often used with ROS packages.  Maybe something like:

```suggestion
``rosdep`` is a dependency management utility that can work with packages and external libraries.
It is most often used with ROS packages, but can also be used outside of ROS.
```

``rosdep`` is a command-line utility for identifying and installing dependencies to build or install a package.
It can be or is invoked when:

- Before building a workspace and needing appropriate dependencies to build the packages within
- Install packages (e.g. ``sudo apt install ros-{DISTRO}-demo-nodes-cpp``) to check the dependencies needed for it to execute

> - and more!

Reviewer 2: This doesn't exactly flow with the opening sentence (`It can be or is invoked when`).

Instead I'll suggest changing this whole block to:

```
It can be or is invoked:

- Before building a workspace and needing appropriate dependencies to build the packages within
- When installing packages (e.g. ``sudo apt install ros-{DISTRO}-demo-nodes-cpp``) to check the dependencies needed for it to execute
- and more!
```

It has the ability to work over a single package or over a directory of packages (e.g. workspace).

> While the name may suggest so, ``rosdep`` is semi-agnostic to ROS.
> You can utilize this powerful tool in non-ROS software project [1]_.
> 
> .. [1] As you may see in this document, ``rosdep`` can be installed as a standalone Python package. To execute it, it relies on the ``rosdep keys`` to be available on the same host. The keys can be downloaded from a public github repo by just a few sinmple commands.
> 

Reviewer 2: ```suggestion
.. note::

    While the name suggests it is for ROS, ``rosdep`` is semi-agnostic to ROS.
    You can utilize this powerful tool in non-ROS software projects by installing it as a standalone Python package.
    Successfully running ``rosdep`` relies on ``rosdep keys`` to be available, which can be downloaded from a public git repository with a few simple commands.
```

Besides that, what are the few simple commands?

A little about package.xml files
--------------------------------

> ``package.xml`` is the file in your software where ``rosdep`` finds the set of dependencies.

Reviewer 2: ```suggestion
The ``package.xml`` is the file in your software where ``rosdep`` finds the set of dependencies.
```

These dependencies are manually populated in the ``package.xml`` file by the package's creators and should be an exhaustive list of any non-builtin libraries and packages it requires.

> You can find more in detail description about the types of dependencies, in e.g. `Catkin's document <http://docs.ros.org/en/melodic/api/catkin/html/howto/format2/catkin_library_dependencies.html>`_.

Reviewer 2: Hm, I really don't want to link from here to the ROS 1 documentation.  Is there another place we can put this?  Maybe import that document into this repository?

Author: Ok...I'm not knowledgeable about this repo's structure yet, but I do want to NOT lose the great bit of info in the referenced page in Catkin. I think that piece of info deserves a new page. For now I remove this line.

Author: Ok, I decided to create a new folder under `Tutorial/Intermediate`, moved rosdep in there, and added another page where I copied the content from Catkin's page I initially referenced to, and modified accordingly. Please re-review.

One of biggest worries for doing this is that this may break the link to the current `Rosdep.rst` page if anything. Hope that is not critical? No idea.

How does rosdep work?
---------------------

``rosdep`` will check for ``package.xml`` files in its path or for a specific package and find the rosdep keys stored within.
These keys are then cross-referenced against a central index to find the appropriate ROS package or software library in various package managers.
Finally, once the packages are found, they are installed and ready to go!

> The central index is known as ``rosdistro``, which `may be found online <https://github.com/ros/rosdistro>`_.

Reviewer 2: I think we should leave this line as-is:

```suggestion
The central index is known as ``rosdistro``, which `may be found here <https://github.com/ros/rosdistro>`_.
```

Author: I wanted to clarify that the "master" of the rosdep key list can be located online, NOT on your host. If "online" doesn't convey the best meaning I'm open for an alternative.

But I'd like to avoid "here". In general, linking text "here" does NOT carry any meaning at all especially in hypertext environment, so I'd avoid it.

We'll explore that more in the next section.

> Actually ``rosdep`` retrieves the central index on to your local host so that it doesn't have to access online every time it runs (On Debian/Ubuntu it's found at ``/etc/ros/rosdep/sources.list.d/20-default.list``, but you really don't need to mess with that at all).

Reviewer 2: ```suggestion
``rosdep`` works by retrieving the central index on to your local host so that it doesn't have to access the network every time it runs (on Debian/Ubuntu the configuration for it is stored in ``/etc/ros/rosdep/sources.list.d/20-default.list``).
```

Actually ``rosdep`` retrieves the central index on to your local host so that it doesn't have to access online every time it runs (On Debian/Ubuntu it's found at ``/etc/ros/rosdep/sources.list.d/20-default.list``, but you really don't need to mess with that at all).

How do I know what keys to put in my package.xml?
-------------------------------------------------

Great question, I'm glad you asked!

For ROS packages (e.g. ``nav2_bt_navigator``), you may simply place the name of the package.

> You can find a list of all released ROS packages in ``rosdistro`` at ``<distro>/distribution.yaml`` for your given ROS distribution.

Reviewer 2: In this section, I think we should keep the bit about `ROS` packages, and then just expand on the non-ROS package.  That is, `rosdep` is still primarily a ROS tool so we should emphasize that.

Author: Hm, in this section I haven't tried to shift the focus toward non-ros. Would you mind giving pointer to sentences that made you make your comment?

How do I use the rosdep tool?
-----------------------------

rosdep installation
~~~~~~~~~~~~~~~~~~~

``apt-get install python3-rosdep`` or ``pip install rosdep``

rosdep operation
~~~~~~~~~~~~~~~~

> 

Reviewer 2: These heading don't exactly flow with the rest of the document.  I'd make a top-level question like `How do I install rosdep?` near the top of the document, and then leave this bit as-is.

`Detailed instructions may be found here <https://github.com/ros/rosdistro/blob/master/CONTRIBUTING.md#rosdep-rules-contributions>`_ for how to contribute new rosdep keys.
If for some reason these may not be contributed openly, it is possible to fork rosdistro and maintain a alternate index for use.

How do I use the rosdep tool?
-----------------------------

rosdep installation
^^^^^^^^^^^^^^^^^^^

If you are using ``rosdep`` with ROS, it is conveniently packaged along with the ROS distribution.
This is the recommended way to get ``rosdep``.
You can install it with:

.. code-block:: bash

>     apt-get install python3-rosdep

Reviewer: I'd suggest adding a note about the `python3-rosdep2` package: it's the upstream Debian/Ubuntu version of `rosdep`, and, unless things have improved recently has the potential to cause all sorts of problems.

Ref: [Missing packages after installing rosdep based on python3-rosdep2 in Noetic](https://answers.ros.org/question/353082) (mentions *Noetic* in the title, but this is `rosdep`, the problem is actually ROS-agnostic) and [python3-rosdep2 install broke ROS](https://answers.ros.org/question/390335).

Reviewer 2: Done in fcdc39ddd74fa8a73f6fbbaca1a71d2aff19712c
